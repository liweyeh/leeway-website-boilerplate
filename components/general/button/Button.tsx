/**
 * Button component
 * @author Liwei Yeh <liwei.yeh@adasasistemas.com>
 */

// Library
import React from 'react';

// CSS Style for this module
import styles from './button.module.scss';

/**
 * Interface for Button props
 */
interface ButtonProps {
	/**
	 * Custom styles of the button
	 */
	customStyles?: string;
	/**
	 * preset type which has its own styling and functionality
	 */
	type?: 'primary' | 'secondary' | 'default' | 'disabled';
	/**
	 * preset size of the button
	 */
	size?: 'small' | 'large';
	/**
	 * Button contents
	 */
	children: React.ReactNode;
	/**
	 * Click handler
	 */
	onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
	/**
	 * aria-label
	 */
	ariaLabel?: string;
}

/**
 * An styled general button
 * @param {ButtonProps} Prop
 * @returns  a button
 */
export const Button = ({
	size,
	type = 'default',
	customStyles,
	children,
	ariaLabel = 'button',
	onClick,
}: ButtonProps) => {
	return (
		<button
			aria-label={ariaLabel}
			type="button"
			className={[styles[`button-${type}`], size && styles[`button-${size}`], customStyles].join(
				' ',
			)}
			onClick={
				type === 'disabled'
					? () => {
							return;
					  }
					: e => {
							onClick(e);
					  }
			}
		>
			{children}
		</button>
	);
};
