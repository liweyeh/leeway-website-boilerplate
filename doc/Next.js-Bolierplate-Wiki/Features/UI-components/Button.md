[[_TOC_]]

# Overview

- Purpose: Pre-styled button with accessbility support
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/button)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 23/08/2022

# Props

\* = required

- children<`React.ReactNode`>\*: Button contents
- onClick<`(e: React.MouseEvent<HTMLButtonElement>) => void`>\*: Click handler
- ariaLabel<`string`>: aria-label
- size<`'small' | 'large'`>: preset size of the button
- type<`'primary' | 'secondary' | 'default' | 'disabled'`>: preset type which has its own styling and functionality
- customStyles<`string`>: Custom styles of the button

# Usage

```tsx
const handleOnClick = e => {
	console.log(e);
};

<Button onClick={handleOnClick} size={'large'}>
	Click Me
</Button>;
```

# Customisation

You can add/change the pre-style by changing SCSS code, such as the following:

```scss
// size
&-small {
	width: 100px;
	min-height: 40px;
}

// styling
&-primary {
	border: none;
	border-radius: var.$border-radius-s;

	@include theme.themify-module(theme.$themes) {
		color: theme.themed('textColor');
		background-color: theme.themed('buttonColor');
		&:hover {
			background-color: theme.themed('buttonHoverColor');
		}
		&:active {
			background-color: theme.themed('buttonActiveColor');
		}
	}

	@include btn();
}
```
