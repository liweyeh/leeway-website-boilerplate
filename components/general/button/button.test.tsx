import { render } from '@testing-library/react';
import { Button } from './Button';
import '@testing-library/jest-dom';

const setup = (type: 'primary' | 'secondary' | 'default' | 'disabled', size: 'small' | 'large') => {
	const { asFragment } = render(
		<Button
			onClick={() => {
				// eslint-disable-next-line no-console
				console.log('button clicked');
			}}
			type={type}
			size={size}
		>
			Button Test
		</Button>,
	);
	return { asFragment };
};

test('test snapshot', () => {
	const { asFragment } = setup('default', 'large');
	const firstRender = asFragment();
	expect(firstRender).toMatchSnapshot();
});
