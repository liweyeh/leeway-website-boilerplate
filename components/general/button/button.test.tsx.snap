// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="button"
    class="button-default button-large "
    type="button"
  >
    Button Test
  </button>
</DocumentFragment>
`;
